core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/io/ChecksumInputStream.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.internal.io;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.zip.Checksum;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * An input-stream that takes a collection of checksums, and updates each checksum when it reads data.
+ */
+@SdkInternalApi
+public class ChecksumInputStream extends FilterInputStream {
+
+    private final Collection<Checksum> checksums = new ArrayList<>();
+
+    public ChecksumInputStream(InputStream stream, Collection<? extends Checksum> checksums) {
```

Comment:
Yep! It takes multiple because multiple "checksums" are computed when a flexible checksum is specified in the signing request - the "flexible"checksum and the sha-256 content hash (or one of the pre-defined constant values, which is why we need the `ConstantChecksum`)

Reply:
Do you have any use cases defined in tests for these variants?

Reply:
Yep. The `ChecksumInputStreamTest` and `FlexibleChecksummerTest` should test multiple checksums being configured.